test/Dialect/FIRRTL/hoist-passthrough.mlir

```diff
@@ -0,0 +1,668 @@
+    firrtl.ref.define %y, %sp_y : !firrtl.probe<uint<5>>
+    firrtl.ref.define %z, %sp_z : !firrtl.probe<uint<5>>
+  }
+}
+
+// -----
+
+// Multiple-instantiation hoisting
+
+// CHECK-LABEL: "MultUTurnTH"
+firrtl.circuit "MultUTurnTH" {
+  // CHECK:      module private @UTurn(in %in: !firrtl.probe<uint<5>>) {
+  // CHECK-NEXT: }
+  firrtl.module private @UTurn(in %in : !firrtl.probe<uint<5>>, out %out : !firrtl.probe<uint<5>>) {
+    firrtl.ref.define %out, %in : !firrtl.probe<uint<5>>
+  }
+  // CHECK: @MultUTurnTH
+  firrtl.module @MultUTurnTH(in %in : !firrtl.uint<5>, in %in2 : !firrtl.uint<5>) {
+    // CHECK: %[[U_IN:.+]] = firrtl.instance
+    %u_in, %u_out = firrtl.instance u @UTurn(in in : !firrtl.probe<uint<5>>, out out : !firrtl.probe<uint<5>>)
+    %ref = firrtl.ref.send %in : !firrtl.uint<5>
+    firrtl.ref.define %u_in, %ref : !firrtl.probe<uint<5>>
+    // CHECK: firrtl.ref.resolve %[[U_IN]]
+    %data = firrtl.ref.resolve %u_out : !firrtl.probe<uint<5>>
+
+    // CHECK: %[[U2_IN:.+]] = firrtl.instance
+    %u2_in, %u2_out = firrtl.instance u2 @UTurn(in in : !firrtl.probe<uint<5>>, out out : !firrtl.probe<uint<5>>)
+    %ref2 = firrtl.ref.send %in2 : !firrtl.uint<5>
+    firrtl.ref.define %u2_in, %ref2 : !firrtl.probe<uint<5>>
+    // CHECK: firrtl.ref.resolve %[[U2_IN]]
+    %data2 = firrtl.ref.resolve %u2_out : !firrtl.probe<uint<5>>
+  }
+}
+
+// -----
+
+// Multiple-level hoist
+
+// CHECK-LABEL: HWMultiLevel
+firrtl.circuit "HWMultiLevel" {
+  // CHECK:      @Child(in %x: !firrtl.uint<5>) {
+  // CHECK-NEXT: }
+  firrtl.module private @Child(in %x: !firrtl.uint<5>, out %y: !firrtl.uint<5>) {
+    firrtl.strictconnect %y, %x : !firrtl.uint<5>
+  }
+  // CHECK:      @Mid(in %x: !firrtl.uint<5>) {
+  // CHECK-NEXT:   %[[C_IN:.+]] = firrtl.instance c
+  // CHECK-NEXT:   firrtl.strictconnect %[[C_IN]], %x
+  // CHECK-NEXT: }
+  firrtl.module private @Mid(in %x: !firrtl.uint<5>, out %y: !firrtl.uint<5>) {
+    %c_x, %c_y = firrtl.instance c interesting_name @Child(in x: !firrtl.uint<5>, out y: !firrtl.uint<5>)
+    firrtl.strictconnect %c_x, %x : !firrtl.uint<5>
+    firrtl.strictconnect %y, %c_y : !firrtl.uint<5>
+  }
+  // CHECK: @HWMultiLevel
+  firrtl.module @HWMultiLevel(in %x: !firrtl.uint<5>, in %x2: !firrtl.uint<5>, out %y: !firrtl.uint<5>) {
+    // CHECK: %[[M_X:.+]] = firrtl.instance m
+    // CHECK: %[[M2_X:.+]] = firrtl.instance m2
+    %m_x, %m_y = firrtl.instance m interesting_name @Mid(in x: !firrtl.uint<5>, out y: !firrtl.uint<5>)
+    %m2_x, %m2_y = firrtl.instance m2 interesting_name @Mid(in x: !firrtl.uint<5>, out y: !firrtl.uint<5>)
+    firrtl.strictconnect %m_x, %x : !firrtl.uint<5>
+    firrtl.strictconnect %m2_x, %x2 : !firrtl.uint<5>
+
+    // CHECK: firrtl.and %[[M_X]], %[[M2_X]]
+    %0 = firrtl.and %m_y, %m2_y : (!firrtl.uint<5>, !firrtl.uint<5>) -> !firrtl.uint<5>
+    firrtl.strictconnect %y, %0 : !firrtl.uint<5>
+  }
+}
+
+// -----
+
+// Don't hoist through public modules
```

Comment:
As mentioned in the comments of the pass, in theory we could hoist out from public modules, as long as we leave the ports as-is.  Leave this alone for now :+1: .